### ISSUE TITLE
Roadmap - what's next?

### ISSUE THREAD

**Issue by jornh** [cefsharp/CefSharp#509]
Kill CefSharp 1 issues, with resolutions:
- convert to being a valid issue for CefSharp 3
- already fixed in CefSharp 3, can't you use that?
- ?

https://github.com/cefsharp/CefSharp/issues?q=is%3Aopen+is%3Aissue+label%3Acef1+-label%3Acef3+

**Comment by sylvain-hamel** [comment]
Hi @jornh @amaitland, there are just 3 github issues linked to milestone 39.0.0. Are theses 3 issues things that work with 37 but that are now broken with 39? 

What else is on the backlog before v39 can be released?

**Comment by amaitland** [comment]
Bit behind in the house keeping, I've updated issues and assigned them to their appropriate milestones.

https://github.com/cefsharp/CefSharp/milestones/37.0.1
https://github.com/cefsharp/CefSharp/milestones/39.0.0

Ideally `39` should include everything for `37.0.1`. We could skip the `37.0.1` release and merge everything into `39.0.0`. That's open for discussion.

If you've got some spare time #718 is of high priority, needs to be resolved before we can release any  `nuget` packages. 

It should also be noted that I rewrote large chunks of the `WPF OSR Rendering` code, so that will need adequate testing before release, so I'd prefer to release at least one more `39-pre` version.

**Comment by amaitland** [comment]
I've resolved #718 

I'm thinking that maybe we do skip `37.0.1` and just move ahead with `39`?

@jornh @jankurianski Thoughts?

**Comment by jornh** [comment]
> I've resolved #718

Yay! Green builds again :smile: :+1: good catch!

Yes I'm all for skipping `37.0.1` and move right to `39`. It's been Chrome stable for a while now and I guess it won't be long before Marshall will branch off CEF for `41` it happened recently with Chromium Dev: https://omahaproxy.appspot.com/

How do you guys see the route towards `39`? 

Looking at the issues currently tagged 37.0.1 and 39 in https://github.com/cefsharp/CefSharp/issues/509#issuecomment-69274810 it looks a bit much if we are aiming for the 28th :stuck_out_tongue_winking_eye: anything critical or that you guys **really** want to see in there?

**Comment by amaitland** [comment]
I think upgrade to `39.0.2171.95` then release another `-pre` version ASAP.

Long as the `WPF` version is stable after the rendering rewrite then I'm happy. I've run the fishbowl stress test quite a few times and everything seems to be going ok. One interesting thing to note is that `CEF` handles all popups, none of our `OSR` popup code is called (which is probably for the best as I don't think it would support more than one from looking at the code).

**Comment by jankurianski** [comment]
Nice detective work on the build error :mag:

As far as CEF is concerned, I feel `39` is better than `37` in terms of performance (already noted by @amaitland) and bug fixes (e.g. you cannot use some weight-specific font family names such as `Segoe UI Light` in `37`, but it works in `39`).

All the issues in the milestones seem to be administrative or the affected users have custom builds with workarounds in place, so I think the majority of users will be happy with a move up in the schedule to the all-around better `39`!

**Comment by amaitland** [comment]
#693 Merged, change log updated, moved issues from `37.0.1` and removed. Just need to update the `appveyor` file so nuget packages are pushed to the primary `MyGet` feed :smile:

**Comment by amaitland** [comment]
`MyGet` packages should be working now :smile: 

I rearranged a few issues on the milestone and closed some issues, so the list for `39` is looking a lot better!

https://github.com/cefsharp/CefSharp/milestones/39.0.0

**Comment by amaitland** [comment]
I think everything is ready for another `-pre` release. I've uploaded `3.2171.1972` packages to the `MyGet` feed. Once we're happy they can all be promoted together :smile: 

https://ci.appveyor.com/project/cefsharp/cefsharp/build/39.0.0-CI638#L19

**Comment by jornh** [comment]
Awesome @amaitland :+1: 

I see the CEF packages are also in the MyGet feed now: https://www.myget.org/gallery/cefsharp. Guess that means we can forget about the staging server completely now?

I can take the packages for a spin with CefSharp.MinimalExample tonight - unless @jankurianski beats me to it :wink:

**Comment by jankurianski** [comment]
I am using the `3.2171.1972 redist` packages from a `master` source build in my app now, and looks good so far.  No time to test the overall packages yet, I will give it a go tomorrow (it will soon be night time here in Sydney).

**Comment by amaitland** [comment]
> Guess that means we can forget about the staging server completely now?

Yeah I think `staging` is redundant now. Nice thing about `MyGet` is they can be deleted and re-uploaded if there's a mistake, so that's a huge bonus from my point of view!

**Comment by amaitland** [comment]
> (it will soon be night time here in Sydney).

Southerners and their crazy daylights savings time :wink: (I'm in Brisbane)

**Comment by jornh** [comment]
Goooodmorning Aussies :sun_with_face: it's 20 past midnight here in the little Kingdom of Denmark now. So just one short semi-incompletely analysed (to put it mildly) observation:

Tested a bit with CI639 al looks fine with WinForms and 32-bit Wpf only with https://www.myget.org/F/cefsharp/api/v2/package/CefSharp.Wpf/39.0.0-CI639 x64 (both Debug and Release) I saw MinimalExample and CS.Wpf.Example crashing.

**steps to reproduce**
X64 Wpf on entering text in the Google.com search box I see an overflow? (As I recall) exception on https://github.com/cefsharp/CefSharp/blob/ec9ad7107d1262883aeac154bfe2ae4eeb861280/CefSharp.Wpf/ChromiumWebBrowser.cs#L628

Maybe regression from 18th dec commit: https://github.com/cefsharp/CefSharp/commit/48c362552360f5ebb8e6246bfb8610dbbf74d8f3 ??

It's entirely possible it's a local problem here, so start seing if it's reproducible at all. 

G'night!

**Comment by amaitland** [comment]
> Goooodmorning Aussies :sun_with_face: it's 20 past midnight here in the little Kingdom of Denmark now. So just one short semi-incompletely analysed (to put it mildly) observation:

Night @jornh! Thanks for the accurate bug report! Uou are indeed correct about the `x64` build, it's the `lParam.ToInt32()` that causes the issue. 

Hopefully commit https://github.com/cefsharp/CefSharp/commit/a250e2d7ed6463ba55a93c43412be4cff5a6f352 resolves the issue. I missed the `LPARAM` being a `64bit` integer when I ported the code over from `cefclient`.

**Comment by amaitland** [comment]
The joys of `C++`! One more change required so the code uses an `int64` in both cases.

https://github.com/cefsharp/CefSharp/commit/0838dcd382cbb0b9d596c5ae261e7e20b751df47

**Comment by amaitland** [comment]
It occurred to me that maybe the following would be cleaner https://github.com/cefsharp/CefSharp/commit/1d7cc92a6a0242268a58567f2b662cec3d66ffbb.

Thoughts?

**Comment by jankurianski** [comment]
> Thoughts?
It does look cleaner, moving the casting down into one line.

**Comment by jankurianski** [comment]
I gave all the `CI646` packages a whirl in `x86` and `x64` with the MinimalExamples - all good!

**Comment by jornh** [comment]
Excellent! Tagged as `v39.0.0-pre02` and awaiting AppVeyor https://ci.appveyor.com/project/cefsharp/cefsharp/build/39.0.0-CI647. Hope the packaging magic with tag detect still works.

Meanwhile I'll give @perlun  a friendly nudge here and via email for the `contributor` rights for @jankurianski

**Comment by amaitland** [comment]
> Hope the packaging magic with tag detect still works

Packages were created successfully. I've pushed them to `nuget.org` :+1:

**Comment by perlun** [comment]
@jornh - I set you up as an admin now. Use your new privileges wisely... :smile:

**Comment by jornh** [comment]
Thanks! Will do/have already done so :smile: by adding @jankurianski.

Welcome aboard Jan!

**Comment by jankurianski** [comment]
Cheers :beers:

I am guessing it is helpful if I review/merge PRs and moderate issues.  Let me know if there's anything else that needs doing :smile:

**Comment by jornh** [comment]
> I am guessing it is helpful if I review/merge PRs and moderate issues. Let me know if there's anything else that needs doing :smile:

Exactly! Continue what you have been doing so well till now in issues and with code, with the addition that now you can also merge and (re-)label issues and a few other things. @amaitland did I forget anything?

I guess next focus is getting `39` marked stuff out the door :beers:

**Comment by amaitland** [comment]
> I am guessing it is helpful if I review/merge PRs and moderate issues. Let me know if there's anything else that needs doing :smile:

@jankurianski Anything you have time for would be greatly appreciated! :smile: 

> @amaitland did I forget anything?

Only thing I can think of is subsequent to merging a PR update the [ChangeLog](https://github.com/cefsharp/CefSharp/wiki/ChangeLog)

**Comment by amaitland** [comment]
On a totally different note, does anyone have comment on coding style? In general I'm happy with everything except `c#` module level variable naming starting with a lowercase, personally I'd prefer to prefix with `_` (same as the `c++` code).

Anyone have thoughts on changing this going forward?

**Comment by jornh** [comment]
> ... subsequent to merging a PR update the ChangeLog

Oh yeah forgot that one!

> except c# module level variable naming starting with a lowercase, personally I'd prefer to prefix with _ (same as the c++ code).

I don't have a strong opinion on it either way as long as we're consistent across our own code base :smile:

**Comment by perlun** [comment]
If I still have a say, please do *not* use _ prefix in C# code. It's bad enough to have it in the C++ code. (I'm not sure if it could be removed there, because it may be that you cannot have a local variable named the same as a class-level variable in that case)

Anyway, please stay away from it whenever possible. IMHO of course. 😎

**Comment by amaitland** [comment]
> Anyway, please stay away from it whenever possible. IMHO of course.

Any particular reason? Whilst most of coding style is personal preference, I beleive this one delivers value (and it's a pretty common standard in `C#` terms, as I'm sure your aware Resharper wants to rename everything by default).

**Comment by amaitland** [comment]
> If I still have a say

I am of course happy to be democratic about it :smile: Hence the question :smile:

**Comment by perlun** [comment]
It looks ugly, and it doesn't deliver enough value. ReSharper can probably tell you anyway whether this is a local variable or class-level - no need to clutter the code with (IMHO) ugly prefixes.

Sent from my iPhone

On 27 jan 2015, at 09:07, Alex Maitland <notifications@github.com<mailto:notifications@github.com>> wrote:


Anyway, please stay away from it whenever possible. IMHO of course.

Any particular reason? Whilst most of coding style is personal preference, I beleive this one delivers value (and it's a pretty common standard in C# terms, as I'm sure your aware Resharper wants to rename everything by default).

-

**Comment by amaitland** [comment]
> It looks ugly, and it doesn't deliver enough value. ReSharper can probably tell you anyway whether this is a local variable or class-level - no need to clutter the code with (IMHO) ugly prefixes.

Reviewing PR's when you only get a small subset of the code I think it has excellent value. `Resharper` isn't always an option :smile: 

Could get into a heated debate, though I don't see that serves any purpose. Happy to take your vote as a :-1: 

Anyone else have comment?

**Comment by perlun** [comment]
Fair point from your end. Anyway, I just think the ugliness isn't worth it.?

Sent from my iPhone

On 28 jan 2015, at 03:14, Alex Maitland <notifications@github.com<mailto:notifications@github.com>> wrote:


It looks ugly, and it doesn't deliver enough value. ReSharper can probably tell you anyway whether this is a local variable or class-level - no need to clutter the code with (IMHO) ugly prefixes.

Reviewing PR's when you only get a small subset of the code I think it has excellent value. Resharper isn't always an option [:smile:]

Could get into a heated debate, though I don't see that serves any purpose. Happy to take your vote as a [:-1:]

Anyone else have comment?

-

**Comment by amaitland** [comment]
I'm planning on releasing `-pre03` either today or tomorrow, anyone have comment?

**Comment by jankurianski** [comment]
Sounds good, lots of fixes in there :+1:

**Comment by jornh** [comment]
Triple  :+1: from me

**Comment by amaitland** [comment]
Release is up on `MyGet` if anyone has time to check the new `-pre03` package that would be much appreciated.

I've created a branch on `MinimalExample` that references the new ones to make life easy :smile: 

https://github.com/cefsharp/CefSharp.MinimalExample/tree/cef/2171

**Comment by jankurianski** [comment]
I tested all 3 `MinimalExample` projects with `-pre03` and they are working well :+1:

**Comment by amaitland** [comment]
Thanks :+1: Everything is looking fine on my end as well so I might push them to `Nuget`

**Comment by amaitland** [comment]
Packages are on `Nuget` :smile:

**Comment by jornh** [comment]
Sweet!  @amaitland I accidentally saw you even deleted it from the MyGet feed so `-pre` don't alphabetically overrule the `-CI` ones :+1:  

In celebration I added back `-pre` in https://github.com/cefsharp/CefSharp/blob/master/README.md

**Comment by amaitland** [comment]
> Sweet! @amaitland I accidentally saw you even deleted it from the MyGet feed so -pre don't alphabetically overrule the -CI ones :+1:

Yeah, had that problem last time, so though it best to delete the `-pre` packages once pushed to `Nuget`

**Comment by amaitland** [comment]
Anyone have any thoughts on how we should inform users of breaking changes? Notes in the change log? Though I'm not sure that anyone actually looks at the change log.

**Comment by jankurianski** [comment]
I think the change log is the best place, as it also shows on the [releases page](https://github.com/cefsharp/CefSharp/releases) (or is that due to you copying it manually?).

We could also give it more exposure on the `README.md` page that everyone arrives at, like this:

***
`...`
# NuGet Packages

Binary releases contain everything needed to embed Chromium in your .Net/CLR application. For usage see [FAQ #8](https://github.com/cefsharp/CefSharp/wiki/Frequently-asked-questions#CefSharp_binaries) item.

Upgrading from an earlier version of CefSharp?  See the [ChangeLog](https://github.com/cefsharp/CefSharp/wiki/ChangeLog) for breaking changes and upgrade tips.

**CI Builds**<br/>
`...`

***

We can add new sections to the ChangeLog for breaking changes.  Then anyone upgrading can scan over these, starting at their old version to the new version, and figure out what they need to fix in their code.  For example:

***
`...`
###### 3.27.0-a0
August 3, 2013, Chromium 27.0.1453.73

- First alpha version of CEF3-based port. Core functionality (normal browsing, back/forward buttons etc) working on WPF. Windows Forms version entirely missing at the moment (not yet ported to CEF3), as well as Javascript integration. [@plundberg]

**Breaking Changes**

- Change from `CEF.Initialize()` to `Cef.Initialize()`

`...`
***

**Comment by amaitland** [comment]
> I think the change log is the best place, as it also shows on the releases page (or is that due to you copying it manually?).

Manually copy the entries from the changelog when creating a release.

Excellent suggestions, let's run with that :+1:

**Comment by amaitland** [comment]
Probably worth going back over the changelog and backdating any breaking changes that are obvious.

**Comment by jankurianski** [comment]
I have added some major renames that would hit Cef1 upgraders to `33.1.0-pre01` (and edited the release notes).

**Comment by amaitland** [comment]
Nice! Hopefully one day `CefSharp1` will just be a distant memory! I'll have a scan over the `ChangeLog` later and see if I can spot anything else.

**Comment by amaitland** [comment]
FYI I've updated to the new `Nuget Restore` method, tests ok locally, `CI` builds.

If anyone has problems then let me know.

https://ci.appveyor.com/project/cefsharp/cefsharp/build/39.0.0-CI751

**Comment by roschler** [comment]
Saw a comment about Windows Phone above.  Did anything ever happen on this front?

**Comment by amaitland** [comment]
> Saw a comment about Windows Phone above. Did anything ever happen on this front?

Comment was about being able to deploy a new version using the web interface, not using `CefSharp` on a phone.

**Comment by roschler** [comment]
Ok, thanks amaitland.  I'm running WebRTC sessions successfully using Windows Forms and CefSharp on the desktop.  I was hoping to do the same on a Windows Phone to establish a video session between a desktop PC and the phone.  I'll keep searching.

**Comment by rassilon** [comment]
I posted to the CefSharp google group asking what folks thought of only supporting VC++ 2013 for CefSharp branches 2272 and later, or whatever CEF branch comes after 2272. I ask because CEF hasn't supported VC++ 2012 for branches >= 1916.

Bill

**Comment by rassilon** [comment]
This isn't meant to suggest a change in the minimum .Net framework version used in CefSharp.

Bill

**Comment by amaitland** [comment]
Cool, let's see what people have to say. When `VS 2015` is released I'd probably look at supporting that. Personally I think having current and one previous is enough.

**Comment by amaitland** [comment]
Rather than keep pushing `39` back as new PR's come through I've decided to bump a few to the `39.0.1` or `41.0.0` milestones.

So what I hope is the final `39.0.0` package is now avaliable on `MyGet`. If anyone has the time to do one last check before we push to `Nuget.org` that would be greatly appreciated.

https://ci.appveyor.com/project/cefsharp/cefsharp/build/39.0.0-CI847

**Comment by amaitland** [comment]
Forgot to update `build.ps1` to reference `3.2171.2069`, so rebuilding packages. If anyone has downloaded, they'll have to clear their nuget local cache.

https://github.com/cefsharp/CefSharp/commit/03e35009c481d9e03e6ae918da79b0b25300d049

**Comment by amaitland** [comment]
Updated `MinimalExample` to use the `MyGet` `39.0.0` packages

https://github.com/cefsharp/CefSharp.MinimalExample

@jornh @rassilon @jankurianski If anyone has 5mins to take them for a spin that would be much appreciated :+1:

**Comment by jankurianski** [comment]
Checked the 3 example projects in `x86` and `x64` and they're all working and show the correct Chrome/CEF version.

**Comment by amaitland** [comment]
@jankurianski Thanks, much appreciated :+1:

**Comment by rassilon** [comment]
I updated to the MyGet 39.0.0 pkgs on my own app, and after having to manually remove the CefSharp/CEF binaries from my Release build directory and rebuild, it seemed to work just fine.

Bill

**Comment by amaitland** [comment]
Excellent, I've tried with `MinimalExample`, with `x86` and `x64` combinations and everything appears to be working fine.

Next step release to `Nuget.org`

**Comment by amaitland** [comment]
Packages are now on `Nuget.org`.

@rassilon Thanks for taking it for a spin :+1:

**Comment by amaitland** [comment]
I've released `39.0.1` to `nuget.org` if anyone is interested. Nothing too existing in this release.

**Comment by amaitland** [comment]
For those interested, the updated plan is to skip `41` and go straight for `43`. The major rewrite happening on the `2272` branch means that there is still quite a bit of work/testing involved still.
